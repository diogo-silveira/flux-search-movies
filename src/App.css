.App {
  background-color: #F7F7F7;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-float infinite 3s ease-in-out;
  }
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

.App-link {
  color: rgb(112, 76, 182);
}

@keyframes App-logo-float {
  0% {
    transform: translateY(0);
  }
  50% {
    transform: translateY(10px)
  }
  100% {
    transform: translateY(0px)
  }
}

.spinner{ 
  content:'';
  display:block;
  position:absolute;
  left:48%;top:40%;
  width:40px;height:40px;
  border-style:solid;
  border-top-color:transparent;
  border-width: 4px;
  border-radius:50%;
}

.cover-spin {
  position:fixed;
  width:100%;
  left:0;right:0;top:0;bottom:0;
  background-color: rgba(255,255,255,0.7);
  z-index:9999;
}

.internet-error {
  height: 30px;
  background: red;
  margin-top: 0;
  font-size: 20px;
}

.internet-error p {
  font-size: 20px;
  line-height: 30px;
  color: #fff;
  margin: 0;
  text-align: center;
}