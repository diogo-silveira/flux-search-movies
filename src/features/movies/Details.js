import React, { useState } from "react";
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';

export default function Details(props){
   const {Title, Actors, Awards, BoxOffice, Country, Poster, Director, Genre, Plot, Production, Runtime, Writer, imdbRating, imdbVotes, Year } = props.item;
    
    return (
        <Modal 
        {...props}
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
        >
        <Modal.Header closeButton>
            <Modal.Title id="contained-modal-title-vcenter">
                {Title}  ({Year})
            </Modal.Title>
        </Modal.Header>
        <Modal.Body>
            <div className="row col-12">
                <div className="col-8"><p>{Plot}</p>
                <p><b>Production</b> - {Production}</p>
                <p><b>Rating</b> - {imdbRating}</p>
                <p><b>Duration</b> - {Runtime}</p>
                <p><b>Votes</b> - {imdbVotes}</p>
                <p><b>Actors</b> - {Actors}</p>
                <p><b>Genre</b> - {Genre}</p>
                <p><b>Country</b> - {Country}</p>
                <p><b>Box Office</b> - {BoxOffice}</p>
                <p><b>Awards</b> - {Awards}</p>
                <p><b>Director</b> - {Director}</p>
                <p><b>Writer</b> - {Writer}</p>
            </div>
                <div className="col-4"><img className="w-100" src={Poster} alt={Title}/></div>
            </div>
        </Modal.Body>
        <Modal.Footer>
            <Button onClick={props.onHide}>Close</Button>
        </Modal.Footer>
    </Modal>
    );

}
