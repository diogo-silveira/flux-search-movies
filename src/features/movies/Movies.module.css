.title {
    height:5rem
}

.picture {
    height:22rem
}

div.highlight:hover {
    cursor: pointer !important;
    background-color: white !important;
    transform: matrix(1, 0, 0, 1, 0, -21);
    transition: width 1s, height 1s, transform 1s;
}