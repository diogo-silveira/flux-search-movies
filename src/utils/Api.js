import axios from "axios";
import { useSelector, useDispatch } from 'react-redux';
import {
    createError,
    displayLoading,
} from '../features/movies/moviesSlice';

const axiosInstance = axios.create({
  baseURL: "http://www.omdbapi.com/?apikey=efdf9830&v=1",
  responseType: "json"
});

axiosInstance.interceptors.request.use(config => {
  console.log('Request was sent');
  return config;
}, error => {
    return Promise.reject(error);
});

axiosInstance.interceptors.response.use((response) => {
  console.log('Response was received');
  return response;
}, error => {
    return Promise.reject(error);
});

export default axiosInstance;