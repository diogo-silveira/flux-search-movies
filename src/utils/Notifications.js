
import React, { useState, Fragment } from 'react';
import { useDispatch } from 'react-redux';
import { Toast } from 'react-bootstrap';
import {
    cleanError
} from '../features/movies/moviesSlice';

export function Notifications(props) {
    const [show, setShow] = useState(true);
    const dispatch = useDispatch();    
    
    const onCloseNotification= () => {
        new Promise((resolve, reject) => resolve(setShow(false)))
                             .then(() => dispatch(cleanError(props.index)));
    }
    const wrapper = React.createRef();

    const notification = (
        <div key={wrapper} style={{position: 'absolute', minHeight: 100, top: props.index ? props.index * 105 : 5, right: 0, zIndex:999999}} ref={wrapper}>
            <Toast key={props.index} onClose={() => onCloseNotification()} show={show} delay={5000} autohide style={{}} >
                    <Toast.Header>
                    <img
                        src="holder.js/20x20?text=%20"
                        className="rounded mr-2"
                        alt=""
                    />
                    <strong className="mr-auto">Error</strong>
                    <small>just now</small>
                    </Toast.Header>
                    <Toast.Body>{ props.message}</Toast.Body>
                </Toast>
        </div>
            );

    return (
        <Fragment>
            { notification }
        </Fragment>
    );
  }