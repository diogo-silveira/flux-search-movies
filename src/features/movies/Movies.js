
import React, { useState, Fragment } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
    selectMoviesList,
    selectMovies,
    selectMovieItem,
    selectMovieDetails,
    cleanMovieItem
} from './moviesSlice';
import style from './Movies.module.css';
import Details from './Details';

export function MovieItem(props) {
    
    const [modalShow, setModalShow] = React.useState(false);
    const dispatch = useDispatch();
    let { imdbID, Title, Poster } = props.movie;
    const divClassHighLight = `${style.highlight} border border-dark bg-white col-lg-2 col-xs-6 col-md-4 col-8 mx-2 my-1 rounded px-0`;    
    const movieItem = useSelector(selectMovieItem);

    const openDetails = () => new Promise((resolve, reject) => resolve(dispatch(selectMovieDetails(imdbID)).then(()=> setModalShow(true))));
    
    const closeDetails = () => {
        setModalShow(false);
        dispatch(cleanMovieItem());
    }

    return(
        <Fragment>
            <div className={ divClassHighLight } key={imdbID} onClick={() => openDetails()}>
                <div className="container">
                        <div className="align-items-center d-flex justify-content-center px-2 py-1 row text-center" style={{ height: "5rem" }}>
                            <span className="w-auto" >{ Title } </span>
                        </div>
                        <div className="row align-items-end" style={{ height: "18rem" }}>
                            <img className="w-100" alt={ Title } src={Poster} style={{ height: 279, width:188}} ></img>                            
                        </div>
                </div>
            </div>
            <Details item={movieItem} show={modalShow} onHide={() => closeDetails() } />         
        </Fragment>
    );
}

export function Movies () {

    const [searchMovie, setSearchMovie] = useState('');
    const [isDisplayMovie, setIsDisplayMovie] = useState(false);
    const dispatch = useDispatch();
    const moviesList = useSelector(selectMoviesList);
   
    const movieDetails = moviesList.map((movie) => <MovieItem key={movie.imdbID} movie={movie}> </MovieItem>);

    const LoadingData = () => <div className="row justify-content-center">
                                { movieDetails }
                              </div>
    

    const DisplayMovies = () => isDisplayMovie ?
                <div className="col-12 h-100 justify-content-center m-0 p-0 row">
                    <div className="col-12 col-lg-10 py-3 row">
                        <LoadingData></LoadingData>
                    </div>        
                </div> : ''
    

    const search = () => new Promise((resolve, reject) => {
            resolve(dispatch(selectMovies(searchMovie)));
        }).then(setIsDisplayMovie(true));

    return (
        <div className="col-12 p-0">
            <div className="col-12 justify-content-center p-0">
                <div className="col-12 justify-content-center pb-4 row pr-0 pt-5 pt-sm-0">
                        <div className="col-12 col-lg-4 col-md-6 pb-2 pr-0">
                            <input className="form-control mr-3" name="searchMovie" type="text" value={ searchMovie } onChange={ e => setSearchMovie(e.target.value)} placeholder="Search a movie..."></input>
                        </div>
                        <div className="col-12 col-lg-2 col-md-4 pr-0">
                            <button name="searchButton" className="btn btn-primary px-5 w-100" onClick={() => search()}>Search</button>
                        </div>
                </div>
            </div>
            <DisplayMovies></DisplayMovies>
        </div>
    )

}