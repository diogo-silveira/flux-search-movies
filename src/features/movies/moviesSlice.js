import { createSlice } from '@reduxjs/toolkit';
import Api from '../../utils/Api';
/* Actions and Reducers */
export const moviesSlice = createSlice({
    name: 'movies',
    initialState: {
        moviesList: [],
        searchMovie: '',
        errors: [],
        loading: false,
        movieItem:{}
    },
    reducers: {
        searchMovie: (state, action) => {
            state.searchMovie = action.payload;
        },
        createMovieList: (state, action) => {
            state.moviesList = action.payload;
        },
        createMovieItem: (state, action) => {
            state.movieItem = action.payload;
        },
        cleanMovieSearch: (state) => {
            state.searchMovie = '';
        },
        cleanMovieItem: (state) => {
            state.movieItem = {};
        },
        createError: (state, action) => {
            console.log(`add- ${action.payload}`)
            state.errors.push(action.payload)
        },
        cleanError: (state, action) => {
            console.log(`remove - ${state.errors}`)
            state.errors = state.errors.splice(action.payload, 1);
        },
        displayLoading: (state, action) => {
            state.loading = action.payload;
        }
    }
});

export const { searchMovie, 
               createMovieItem, 
               createMovieList, 
               createError, 
               displayLoading,
               cleanError, 
               cleanMovieItem } = moviesSlice.actions;

/* Methods */
export const selectMovies = movie => dispatch => {
    return Api.get(`&s=${movie}`)
        .then(({data}) => data !== undefined ? data.Search : undefined)
        .then((data) => {
            if(data === undefined)
                return dispatch(createError(`The search with ${movie} didn't find any results.`))

            dispatch(createMovieList(data))
        })
        .catch((error) => dispatch(createError(error.message)));
}

export const selectMovieDetails = id => dispatch => {
    return Api.get(`&i=${id}`)
    .then(({data}) => dispatch(createMovieItem(data)))
    .catch((error) => dispatch(createError(error.message)));
}


/* Selectors */
export const selectMoviesList = state => state.movies.moviesList;
export const selectMovieItem = state => state.movies.movieItem;
export const selectErrors = state => state.movies.errors;
export const selectLoading = state => state.loading;

export default moviesSlice.reducer;
