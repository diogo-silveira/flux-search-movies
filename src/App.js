import React from 'react';
import { Movies } from './features/movies/Movies'
import { useSelector, useDispatch } from 'react-redux';
import './App.css';
import { Notifications } from './app/Notifications'
import { selectErrors } from './features/movies/moviesSlice';

function App() {
  
  const selectError = useSelector(selectErrors);
  const notifications = selectError.map((error, index) => <Notifications index={ index } message={ error }></Notifications>);

  const logo = 'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQM_62ZYb6Q812SZfSZscOi9-7HtBYWH2P4YlcQJ3Sob_4QKcQY&usqp=CAU';
  return (
    <div className="App h-100 position-relative">
       { notifications }
      <div className="col-12 h-100 position-relative px-0">
        <div className="col-12 h-auto text-center">
          <img src={logo} className="App-logo" alt="logo" />
        </div>
          <Movies />
      </div>
     
    </div>
  );
}

export default App;
