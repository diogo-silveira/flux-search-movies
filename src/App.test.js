import React from 'react';
import { Provider } from 'react-redux';
import store from './app/store';
import App from './App';
import { shallow, render, mount   } from 'enzyme';
import 'enzyme';
import { Movies } from './features/movies/Movies'
import { Notifications } from './utils/Notifications'

describe('Test App', () => {

  test('render the app', () => {
    const component = shallow(<Provider store={store}> <App /> </Provider>);
    expect(component).toMatchSnapshot();
  });

  test('should have logo', () => {
    const component = mount( <Provider store={store}> <App /> </Provider>);
    expect(component.find('#logo')).toBeDefined();
    expect(component.find('GlobalLoader')).toBeDefined();
    component.unmount();
  });
  
});

describe('render Movies', () => {
  test('Movies', () => {
    const movie = shallow(<Provider store={store}><Movies/></Provider>);
    expect(movie).toMatchSnapshot();
  })
})

describe('render Notifications', () => {
  test('Notifications', () => {
    const movie = shallow(<Provider store={store}><Notifications/></Provider>);
    expect(movie).toMatchSnapshot();
  })
})
