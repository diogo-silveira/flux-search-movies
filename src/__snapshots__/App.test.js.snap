// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App render the app 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
   
  <App />
   
</ContextProvider>
`;

exports[`render Movies Movies 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Movies />
</ContextProvider>
`;

exports[`render Notifications Notifications 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Notifications />
</ContextProvider>
`;
